### DISCUSSION TITLE
Off-board crazyflie controller

### DISCUSSION THREAD

**Discussion by ntukenmez3** [bitcraze/discussions#993] [Q&A] [answered]
I would like to ask some questions about cf2 controller. To be clear, is it possible to use an off-board controller for cf2, my mind has been quite confused as I read the bitcraze website, especially "Adding an estimator or controller" section (i.e.,  [adding a controller on crazyflie](https://www.bitcraze.io/2023/02/adding-an-estimator-or-controller/)). 
As much as I understood from that page, we can configure or change the onboard controller existing on cf2 (via C language).
I'm curious about adding an off-board controller (via python), I mean a separate controller. Could you help me about that, we need to make some progress on the controller design for cf2, thanks.

**Comment by krichardsson** [comment, marked as answer]
I think it is doable to write off-board controllers (I have not tried), but you have to take latency into account and I don't think it is trivial. You might have to split your controller so that the gyro stabilization still runs on the CF (as off-board might not be fast enough), while higher levels of control runs off-board. Unfortunately there are no APIs or easy way to do this and you will have to dig around in the code.

> Could you help me about that, we need to make some progress on the controller design for cf2, thanks.

First you need to transfer selected parts of the [estimator state](https://github.com/bitcraze/crazyflie-firmware/blob/7ae404b451f73a23beaf65bc0806ff6811862f07/src/modules/src/stabilizer.c#L749-L815) to your PC, you can possibly use the [logging system](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/logparam/#logging) for this or [create your own packet](https://github.com/bitcraze/crazyflie-firmware/blob/master/examples/app_appchannel_test/README.md).
When the controller has crunched one iteration of data you need to transfer the result to the CF, the best way is probably to use a custom packet (you can find a related discussion in #184). You probably don't want to control the motors directly, but instead communicate the data in the [`control_t`](https://github.com/bitcraze/crazyflie-firmware/blob/7ae404b451f73a23beaf65bc0806ff6811862f07/src/modules/interface/stabilizer_types.h#L192-L222) type and [hook in to the stabilzer loop](https://github.com/bitcraze/crazyflie-firmware/blob/7ae404b451f73a23beaf65bc0806ff6811862f07/src/modules/src/stabilizer.c#L319-L327) somehow. 

As I have not done this myself the above might not work, but this is where I would start to poke around. 

I also answered [your other post](#413) ([please avoid double posting](#55))

**Reply by ntukenmez3** [reply]
Thank you for your helpful explanation. To be clear, we're trying to implement a position tracking controller. I think the latency might be reasonable since this is the last control loop on crazyflie (am I right?) as in below. ![image](https://github.com/bitcraze/discussions/assets/143043483/8abcc49f-3bf8-4c0e-ae10-a013e639ff44)
However, my mind is not clear about one off-board controller and one on-board controller for same position tracking control loop. Just as a recommendation, it would be helpful to provide a trivial controller package to test any controller.
